mitiq/mitiq_pyquil/compiler.py

```diff
@@ -0,0 +1,343 @@
+        (inst "CNOT" ()                            p q)
+        (inst "RZ"   (list (param-* alpha -0.5d0)) q)
+        (inst "CNOT" ()                            p q)
+        (inst "RZ"   (list (param-* alpha  0.5d0)) q)
+        (inst "RZ"   (list (param-* alpha  0.5d0)) p))
+    """
+    p = Program()
+    p.inst(_CNOT(q0, q1))
+    p.inst(RZ(-0.5 * angle, q1))
+    p.inst(_CNOT(q0, q1))
+    p.inst(RZ(0.5 * angle, q1))
+    p.inst(RZ(0.5 * angle, q0))
+    return p
+
+
+def _H(q: int) -> Program:
+    """
+    A Hadamard in terms of RX(+-pi/2) and RZ(theta)
+
+    .. note:
+        This introduces a different global phase! Don't control this gate.
+    """
+    p = Program()
+    p.inst(_RY(-np.pi / 2, q))
+    p.inst(RZ(np.pi, q))
+    return p
+
+
+def _ISWAP(q0: int, q1: int) -> Program:
+    """
+    An ISWAP as an XY(pi). Of course, assumes XY is available.
+    """
+    p = Program()
+    p += XY(np.pi, q0, q1)
+    return p
+
+
+def _PHASE(angle: float, q: int) -> Program:
+    """
+    from quilc:
+
+    (define-compiler PHASE-to-RZ ((phase-gate ("PHASE" (alpha) q)))
+       (inst "RZ" `(,alpha) q))
+    """
+    p = Program()
+    p += RZ(angle, q)
+    return p
+
+
+def _RX(angle: float, q: int) -> Program:
+    """
+    A RX in terms of native RX(+-pi/2) and RZ gates.
+    """
+    p = Program()
+    p += RZ(pi / 2, q)
+    p += RX(pi / 2, q)
+    p += RZ(angle, q)
+    p += RX(-pi / 2, q)
+    p += RZ(-pi / 2, q)
+    return p
+
+
+def _RY(angle: float, q: int) -> Program:
+    """
+    A RY in terms of RX(+-pi/2) and RZ(theta)
+    """
+    p = Program()
+    p += RX(pi / 2, q)
+    p += RZ(angle, q)
+    p += RX(-pi / 2, q)
+    return p
+
+
+def _S(q: int) -> Program:
+    """
+    An S in terms of RZ(theta)
+    """
+    return Program(RZ(np.pi / 2, q))
+
+
+def _SWAP(q0: int, q1: int) -> Program:
+    """
+    A SWAP in terms of _CNOT
+
+    .. note:
+        This uses :py:func:`_CNOT`, so it picks up a global phase.
+        Don't control this gate.
+    """
+    p = Program()
+    p.inst(_CNOT(q0, q1))
+    p.inst(_CNOT(q1, q0))
+    p.inst(_CNOT(q0, q1))
+    return p
+
+
+def _T(q: int, dagger: bool = False) -> Program:
+    """
+    A T in terms of RZ(theta)
+    """
+    if dagger:
+        return Program(RZ(-np.pi / 4, q))
+    else:
+        return Program(RZ(np.pi / 4, q))
+
+
+def _X(q: int) -> Program:
+    """
+    An X in terms of RX(pi)
+
+    .. note:
+        This introduces a global phase! Don't control this gate.
+    """
+    p = Program()
+    p += RX(np.pi, q)
+    return p
+
+
+def _Y(q: int) -> Program:
+    """
+    A Y in terms of _RY
+    """
+    p = Program()
+    p += _RY(np.pi, q)
+    return p
+
+
+def _Z(q: int) -> Program:
+    """
+    A Z in terms of RZ
+    """
+    p = Program()
+    p += RZ(np.pi, q)
+    return p
+
+
+# This function is taken from cirq. License: apache 2.
+def match_global_phase(
+    a: np.ndarray, b: np.ndarray
+) -> Tuple[np.ndarray, np.ndarray]:
```

Comment:
Given that we depend on cirq, could we just import this?

Reply:
+1.

Reply:
Good point!